fake_tensor.py

```diff
@@ -0,0 +1,171 @@
+aten = torch.ops.aten
+
+
+# Meta tensors give you the ability to run PyTorch code without having to
+# actually do computation through tensors allocated on a `meta` device.
+# Because the device is `meta`, meta tensors do not model device propagation.
+# FakeTensor extends MetaTensors to also carry an additional `fake_device`
+# which tracks devices that would have been used.
+
+
+class FakeTensor(BaseTensor):
+    fake_device: torch.device
+
+    @staticmethod
+    def __new__(cls, elem, device):
+        return super().__new__(cls, elem)
+
+    def __init__(self, elem, device: Union[torch.device, str]):
+        # elem does not need to be recorded, because FakeTensor *is a* elem
+        assert elem.device.type == "meta"
+        device if isinstance(device, torch.device) else torch.device(device)
+        assert device.type != "meta"
+        self.fake_device = device
+
+    @staticmethod
+    def from_tensor(t):
+        existing_device = t.device
+        return FakeTensor(t.to(device="meta"), existing_device)
+
+    @property
+    def device(self):
+        return self.fake_device
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        # Run the original computation
+        r = super().__torch_dispatch__(func, types, args, kwargs)
+        kwargs = kwargs if kwargs else {}
+
+        def wrap(e, device):
+            # inplace ops can return fake tensors
+            if isinstance(e, torch.Tensor) and not isinstance(e, cls):
+                return FakeTensor(e, device)
+            else:
+                return e
+
+        # Pytorch device is kwarg-only, except for _pin_memory/pin_memory
+        # which is not yet supported by meta-tensors
+        assert (
+            func != aten._pin_memory.default and func != aten.pin_memory.default
+        ), f"NYI: {func}"
+
+        # if device is specified, use that
+        # not sure this is actually needed.. device only shows up in constructors
+        if kwargs.get("device", None):
+            return tree_map(partial(wrap, device=kwargs["device"]), r)
+
+        # operators which copy size from another tensor do not
+        # also take device from the size tensor
+        # other size_as operators are not builtin operators
+        if func == torch.ops.aten.resize_as_.default:
+            # TODO: https://github.com/pytorch/pytorch/pull/77182
+            _, new_kwargs = normalize_function(
+                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+            )
+            # device of the input is returned
+            return tree_map(partial(wrap, device=new_kwargs["input"].device), r)
+
+        # TODO, file issue, this path does not get exercised
+        if func is torch.ops.aten.type_as:
+            _, new_kwargs = normalize_function(
+                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+            )
+            return tree_map(partial(wrap, device=new_kwargs["other"].device), r)
```

Comment:
It's a composite, that's why it doesn't get called

```
- func: type_as(Tensor self, Tensor other) -> Tensor
  variants: method
```

Reply:
I guess I just need to define `__torch_function__` to deal with some of the composite operators

Reply:
Why? Shouldn't they just be properly handled by the natural decomposition?

Reply:
Because `meta_tensor.type_as(meta_tensor)` short-circuits, they both have the same type

Reply:
I will probably also need to handle all variants of aten::to ..

Reply:
Do you have any details on this ?

Reply:
Work like https://github.com/pytorch/pytorch/pull/77396 should allow us to intercept all the calls on Tensors. And in particular avoid this short circuit?